ui: Alert Refactoring, remove most Banner Alerts

mrtracy opened the pull request:
This is a significant refactoring of the systems previously used to display
banner-type notifications.

+ All logic on whether or not to display alerts has been moved to a series of
selectors in `redux/alerts.ts`. After many factoring attempts, using a selector
for each alert seemed like the appropriate design based on current requirements.
+ A new sync service has been added in `alerts.ts`, which is responsible for
loading data from the server which is needed to display alerts. This was chosen
because loading this data using the same method as other components (e.g. how
metric data is loaded) was proving tedious given the nature of alert-relevant
information (only needs to be loaded once, a variety of different settings).
+ Most alerts have been moved from Banners into a dismissable panel which is
displayed on the cluster page. This makes them significantly less disruptive.
+ Significant testing has been added for the alert display logic and the alert
sync service.

Not in this issue:

+ The "cluster disconnected" banner has not been modified, it still displays at
its previous position. The styling of this banner will be changed in a
subsequent pull request, but it will remain visible across all pages because it
is relevant to all pages.

Additional Refactoring changes:

+ Renamed `AdminUIState.ui` -> `AdminUIState.localSettings` to make its purpose
more clear.

Images of new alerts (all three levels on display, although in this example the "Help Us" notification has been artificially changed to a critical-level warning for the purposes of visualizing it):

![screen shot 2017-04-12 at 7 45 09 pm](https://cloud.githubusercontent.com/assets/6969858/25024787/58ee7f7c-206d-11e7-8021-49e30086333d.png)

Commit:
ui: Alert Refactoring, remove most Banner Alerts

This is a significant refactoring of the systems previously used to display
banner-type notifications.

+ All logic on whether or not to display alerts has been moved to a series of
selectors in `redux/alerts.ts`. After many factoring attempts, using a selector
for each alert seemed like the appropriate design based on current requirements.
+ A new sync service has been added in `alerts.ts`, which is responsible for
loading data from the server which is needed to display alerts. This was chosen
because loading this data using the same method as other components (e.g. how
metric data is loaded) was proving tedious given the nature of alert-relevant
information (only needs to be loaded once, a variety of different settings).
+ Most alerts have been moved from Banners into a dismissable panel which is
displayed on the cluster page. This makes them significantly less disruptive.
+ Significant testing has been added for the alert display logic and the alert
sync service.

Not in this issue:

+ The "cluster disconnected" banner has not been modified, it still displays at
its previous position. The styling of this banner will be changed in a
subsequent pull request, but it will remain visible across all pages because it
is relevant to all pages.

Additional Refactoring changes:

+ Renamed `AdminUIState.ui` -> `AdminUIState.localSettings` to make its purpose
more clear.

Resolves #12890
Resolves #12590

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14915)
<!-- Reviewable:end -->

tamird commented:
It's unfortunate that there's so much duplication in the alerts, and their tests. Is there no way to share some of the base functionality?

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> though did not review html/css thoroughly (did not look at css at all).

---

Reviewed 25 of 28 files at r1.
Review status: 25 of 28 files reviewed at latest revision, 26 unresolved discussions, some commit checks failed.

---

*[pkg/ui/src/index.tsx, line 143 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheBEGaePOagHYWEF4l:-KheBEGaePOagHYWEF4m:b-jriggh) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/index.tsx#L143)):*
> ```XML
> );
> 
> // Subscribe store listeners and sync services.
> ```

remove, not helpful!

---

*[pkg/ui/src/containers/nodeGraphs.tsx, line 620 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheDlkylTCszmCGJ6W6:-KheDlkylTCszmCGJ6W7:b-ag9195) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/containers/nodeGraphs.tsx#L620)):*
> ```XML
>       </div>
>       <div className="l-columns__right">
>         <Alerts />
> ```

so this is the "cluster" page? odd that the file name doesn't match.

---

*[pkg/ui/src/containers/banner/bannerContainer.tsx, line 5 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheBYh3n2smQaOpuPeN:-KheBYh46L73XBQgf8t8:b-kqdvfa) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/containers/banner/bannerContainer.tsx#L5)):*
> ```XML
> 
> /**
>  * This is the outer component for all banners
> ```

this lies now, yes?

---

*[pkg/ui/src/redux/alerts.spec.ts, line 24 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-Khi-8G1_1Tadc6jSXhe:-Khi-8G1_1Tadc6jSXhf:b-61e3s0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.spec.ts#L24)):*
> ```XML
>   let state: typeof store.getState;
> 
>   beforeEach(function () {
> ```

not a big fan of this pattern. just sayin'

---

*[pkg/ui/src/redux/alerts.spec.ts, line 57 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-Khi-MmLg43o_dF6njKY:-Khi-MmLg43o_dF6njKZ:b-1x2ry0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.spec.ts#L57)):*
> ```XML
> 
>       it("doesn't display when opt-in option set", function () {
>         dispatch(setUIDataKey(KEY_HELPUS, null));
> ```

what's the point of this line?

---

*[pkg/ui/src/redux/alerts.spec.ts, line 82 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-Khi-aOzmk6V11hctsrp:-Khi-aOzmk6V11hctsrq:b-oqxr2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.spec.ts#L82)):*
> ```XML
>         // We know that this particular alert returns a thunk (saving persistent
>         // UI data), and that the thunk returns a Promise<void>.
>         let p1: Promise<void> = dispatch(alert.dismiss);
> ```

const. help me understand this type annotation; what is the inferred type without it? why can't you `dispatch().then` directly?

---

*[pkg/ui/src/redux/alerts.ts, line 31 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheEpw0xQFYGWe2dbMy:-KheEpw1c3t1v9XFzWDJ:b-gtyqke) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L31)):*
> ```XML
>   // Title to display with the alert.
>   title: string;
>   // Function that returns the text of this alert.
> ```

function?

---

*[pkg/ui/src/redux/alerts.ts, line 35 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheEsZbAQA6Yeqy5nHt:-KheEsZbAQA6Yeqy5nHu:b-ok5n9s) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L35)):*
> ```XML
>   // Optional hypertext link when clicking alert.
>   link?: string;
>   // ThunkAction which will result in this alert being dismissed. This
> ```

? move it down to Alert, I think.

---

*[pkg/ui/src/redux/alerts.ts, line 40 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheF-UNGdA245qn5Q16:-KheF-UOpz7zcMhQF_JM:b4scxkw) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L40)):*
> ```XML
> 
> export interface Alert extends AlertInfo {
>   dismiss: ThunkAction<any, AdminUIState, any>;
> ```

what are these `any`s?

---

*[pkg/ui/src/redux/alerts.ts, line 53 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheFDbCTH5ePmWvIqUx:-KheFDbCTH5ePmWvIqUy:b-v0twh3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L53)):*
> ```XML
> export const optinAttributesSelector = createSelector(
>   (state: AdminUIState) => state.uiData,
>   (uiData) => uiData && uiData[KEY_HELPUS] && uiData[KEY_HELPUS].data as OptInAttributes,
> ```

_.has?

---

*[pkg/ui/src/redux/alerts.ts, line 58 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-Khh_sTswYkkFc55q5Ka:-Khh_sTswYkkFc55q5Kb:ba413is) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L58)):*
> ```XML
> // optinAttributesLoaded is a boolean that indicates whether the optinAttributes
> // have been loaded yet.
> // TODO(mrtracy): Refactor so that we can distinguish "never loaded" from
> ```

why is this needed in addition to the above selector?

---

*[pkg/ui/src/redux/alerts.ts, line 75 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-Khh_zPGgSaC2amTEPJI:-Khh_zPGgSaC2amTEPJJ:bfq9dsu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L75)):*
> ```XML
>   (optinAttributes, attributesLoaded, helpusBannerDismissed): Alert => {
>     if (helpusBannerDismissed) {
>       return undefined;
> ```

is this allowed in the presence of strict null checking?

---

*[pkg/ui/src/redux/alerts.ts, line 84 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KhhaHxjx7dgPI6hWXOA:-KhhaHxkdM4bvvMPkrOY:b-i6zl1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L84)):*
> ```XML
>     // If server-side opt-in settings indicate the user has already explicitly
>     // set an opt-in preference, do not display the notification.
>     // TODO(mrtracy): Log an issue to create a time-limited persistent
> ```

just file the issue, this TODO will be lost.

---

*[pkg/ui/src/redux/alerts.ts, line 100 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KhhaTu5V8byArdAweaf:-KhhaTu5V8byArdAweag:bgb9qky) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L100)):*
> ```XML
>         // Dismiss with persistence on server.
>         const newAttributes = optinAttributes ? _.clone(optinAttributes) : new OptInAttributes();
>         // "dismissed" counts the number of times this banner has been
> ```

mentioned this elsewhere, let's go ahead and simplify this now? being able to do this refactor confidently is the reason we use TS.

---

*[pkg/ui/src/redux/alerts.ts, line 121 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-Khhaeau8QEFTy44keiF:-Khhaeau8QEFTy44keiG:bmh5xhb) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L121)):*
> ```XML
> export const versionsSelector = createSelector(
>   nodeStatusesSelector,
>   (nodeStatuses) => nodeStatuses && _.uniq(_.map(nodeStatuses, (status) => status.build_info && status.build_info.tag)),
> ```

is this nil guard necessary? is build info nullable?

---

*[pkg/ui/src/redux/alerts.ts, line 132 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KhhalHQf3_soElL4OCY:-KhhalHQf3_soElL4OCZ:bq15dj9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L132)):*
> ```XML
>   (versions, versionMismatchDismissed): Alert => {
>     if (versionMismatchDismissed) {
>       return undefined;
> ```

same question about strict null checks

---

*[pkg/ui/src/redux/alerts.ts, line 135 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KhhanbkuKHu6KnKq9uq:-KhhanbkuKHu6KnKq9ur:b1iqols) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L135)):*
> ```XML
>     }
> 
>     if (!versions || versions.length <= 1) {
> ```

how can versions be null or undefined?

---

*[pkg/ui/src/redux/alerts.ts, line 151 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KhhaxFBuvwVeURsXjQJ:-KhhaxFBuvwVeURsXjQK:b-z8e3d9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L151)):*
> ```XML
>   });
> 
> const newVersionDismissedPersistentLoadedSelector = createSelector(
> ```

same question about why this is needed in addition to the below, and also this probably needs the same TODO as the above corollary.

---

*[pkg/ui/src/redux/alerts.ts, line 183 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-Khhb9588oUj8cdSVeQU:-Khhb9588oUj8cdSVeQV:b63gwsn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L183)):*
> ```XML
> 
>     // Check persistent dismissal, also valid for one day.
>     if (!newVersionDismissedPersistentLoaded || newVersionDismissedPersistent.isAfter(yesterday)) {
> ```

sketchy null checks i think

---

*[pkg/ui/src/redux/alerts.ts, line 199 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KhhbCx-afv3mPizQX8T:-KhhbCx-afv3mPizQX8U:b-dvrg80) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L199)):*
> ```XML
>       dismiss: (dispatch) => {
>         const dismissedAt = moment();
>         // Dismiss locally.
> ```

these repeat a bunch, are they useful? i'd remove them all.

---

*[pkg/ui/src/redux/alerts.ts, line 227 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KhhyOSpnd8hydS1gOGn:-KhhyOSpnd8hydS1gOGo:b7vb4t7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L227)):*
> ```XML
>   versionsSelector,
>   (builds) => {
>     if (!builds || builds.length !== 1) {
> ```

I think you can `return _.head(builds);`

---

*[pkg/ui/src/redux/alerts.ts, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KhhyY9m-A87X4aG3gDA:-KhhyY9oG9Q1nC_jiyJu:bo1ejf9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L245)):*
> ```XML
>   // Memoizers to prevent unnecessary dispatches of alertDataSync if store
>   // hasn't changed in an interesting way.
>   let lastUIData: any;
> ```

this can do better than any, can't it? at least, `typeof AdminUIState.uiData`.

---

*[pkg/ui/src/redux/alerts.ts, line 256 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-Khhzkz0nRX4Ha-FJdPB:-Khhzkz1I-9-V9M8LQ5p:b8scwml) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L256)):*
> ```XML
>       const keysToMaybeLoad = [KEY_HELPUS, VERSION_DISMISSED_KEY];
>       const keysToLoad = _.filter(keysToMaybeLoad, (key) => {
>         if (_.has(uiData, key) || isInFlight(state, key)) {
> ```

`return !(_.has(uiData, key) || isInFlight(state, key))`?

---

*[pkg/ui/src/redux/apiReducers.ts, line 29 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheDK3ZSF9kfBu6gxIc:-KheDK3ZSF9kfBu6gxId:b-v1xhve) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/apiReducers.ts#L29)):*
> ```XML
> 
> export const nodesReducerObj = new CachedDataReducer((req: api.NodesRequestMessage, timeout?: moment.Duration) => api.getNodes(req, timeout).then(rollupStoreMetrics), "nodes", moment.duration(10, "s"));
> export const refreshNodes = nodesReducerObj.refresh;
> ```

remove this now that the whole reducer is exported?

---

*[pkg/ui/src/redux/apiReducers.ts, line 35 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheDOJ0SpUMwOj-jgyD:-KheDOJ0SpUMwOj-jgyE:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/apiReducers.ts#L35)):*
> ```XML
> 
> export const versionReducerObj = new CachedDataReducer(versionCheck, "version");
> export const refreshVersion = versionReducerObj.refresh;
> ```

ditto

---

*[pkg/ui/src/redux/uiData.ts, line 34 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheBsJ9KkHQPS90FQ8k:-KheBsJ9KkHQPS90FQ8l:b-4vemx4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/uiData.ts#L34)):*
> ```XML
>   optin: boolean = null; // Did the user opt in/out of reporting usage
>   /**
>    * Number of times the user has dismissed the opt-in banner. This was made a
> ```

any reason not to make it a boolean now?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-:-Khi0Mi008hyKwXu4lBX:bsojp84)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
I spent a lot of mental effort to come up with a de-duplication pattern, but nothing was forthcoming - all of the current alerts have slightly different dismissal criteria. We will be getting rid of the help-us alert, so when we take another crack at this later we can probably get it a bit more compact; however, for the time being I went with the factoring you see here, where the duplicated parts are at least small.

---

Review status: 20 of 28 files reviewed at latest revision, 26 unresolved discussions.

---

*[pkg/ui/src/index.tsx, line 143 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheBEGaePOagHYWEF4l:-Khu14wxWwj4kAAd0MRz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/index.tsx#L143)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

remove, not helpful!
</blockquote></details>

Done.

---

*[pkg/ui/src/containers/nodeGraphs.tsx, line 620 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheDlkylTCszmCGJ6W6:-Khu16PUk62FkY6xfcab:bfl3kwu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/containers/nodeGraphs.tsx#L620)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

so this is the "cluster" page? odd that the file name doesn't match.
</blockquote></details>

Yes, the file names are awful right now, I have it on my list to refactor them into something that makes sense. It will likely be one of the first things I do after 1.0.

---

*[pkg/ui/src/containers/banner/bannerContainer.tsx, line 5 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheBYh3n2smQaOpuPeN:-Khu1SkIetL7Wevw6Blz:biy6sta) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/containers/banner/bannerContainer.tsx#L5)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this lies now, yes?
</blockquote></details>

Nope, the new alerts aren't banners, they're panels. This is the only banner.

This is also getting deleted in an upcoming PR, but no, it doesn't lie.

---

*[pkg/ui/src/redux/alerts.spec.ts, line 24 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-Khi-8G1_1Tadc6jSXhe:-Khu2Khaa5LiZenRdVwJ:b-w8mlee) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.spec.ts#L24)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

not a big fan of this pattern. just sayin'
</blockquote></details>

While it may make it slightly harder to read tests, it certainly makes it faster to write tests, especially when you have so many small, similar cases. When its in the same file as all of the tests it effects, my opinion is that it's a fine tradeoff.

---

*[pkg/ui/src/redux/alerts.spec.ts, line 57 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-Khi-MmLg43o_dF6njKY:-Khu1aOxixUq7ya58H0N:buoephi) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.spec.ts#L57)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what's the point of this line?
</blockquote></details>

This is a confusing problem with UI Data, we need to distinguish "user has no setting on the server" from "we never even checked the server for this setting".  We detect this using _.has() on the UI Data object, rather than using more readable logic.

---

*[pkg/ui/src/redux/alerts.spec.ts, line 82 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-Khi-aOzmk6V11hctsrp:-Khu2BHukfzjjnHGfyev:b-86ug3t) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.spec.ts#L82)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

const. help me understand this type annotation; what is the inferred type without it? why can't you `dispatch().then` directly?
</blockquote></details>

Ah, there was no reason, just copied and pasted from another test that was using two promises and then a `Promise.All`.  Changed to use your suggestion.

---

*[pkg/ui/src/redux/alerts.ts, line 31 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheEpw0xQFYGWe2dbMy:-Khu5THBPYNQiTc0rFj1:beoaa7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L31)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

function?
</blockquote></details>

Ah, see, I was playing around with a lot of factorings.  Fixed.

---

*[pkg/ui/src/redux/alerts.ts, line 35 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheEsZbAQA6Yeqy5nHt:-Khu5T-JOjEtVUGtQ2KN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L35)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

? move it down to Alert, I think.
</blockquote></details>

Done.

---

*[pkg/ui/src/redux/alerts.ts, line 40 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheF-UNGdA245qn5Q16:-Khu3aYDJrt0reM4PeY1:bw4afz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L40)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what are these `any`s?
</blockquote></details>

The first is the return type of the Thunk, which is what we use in tests to synchronize with asynchronous actions.  Since you brought attention to this, I made this type "Promise<void>", and now all of the dismiss functions return one (previously, two out of three returned such a promise).

The second is some weird "extraArguments" that we don't use, so i've just set it to `void`.

---

*[pkg/ui/src/redux/alerts.ts, line 53 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheFDbCTH5ePmWvIqUx:-Khu5f72vB5e4WGD_eoh:be80jtv) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L53)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

_.has?
</blockquote></details>

That will return true if uiData[KEY_HELPUS] is null, but not undefined.

---

*[pkg/ui/src/redux/alerts.ts, line 58 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-Khh_sTswYkkFc55q5Ka:-Khu5qraWNXzegwhm-Mz:b-dpohbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L58)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why is this needed in addition to the above selector?
</blockquote></details>

This keeps the ugly logic explicit.  I suppose we could put it in the notification selector, but to be honest this is getting ripped out before 1.0 so I'm gonna leave it as is.

---

*[pkg/ui/src/redux/alerts.ts, line 75 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-Khh_zPGgSaC2amTEPJI:-Khu68k9qkxMhdGz5GJR:bwccqlc) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L75)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is this allowed in the presence of strict null checking?
</blockquote></details>

I think the function would just have to return Alert | undefined, we could easily switch all of these to null.  Is that what you mean?

---

*[pkg/ui/src/redux/alerts.ts, line 84 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KhhaHxjx7dgPI6hWXOA:-Khu6c0G_L6MWQ_cJhqG:b-ozxim7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L84)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

just file the issue, this TODO will be lost.
</blockquote></details>

This is getting ripped out anyway, just gonna delete the todo.

---

*[pkg/ui/src/redux/alerts.ts, line 100 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KhhaTu5V8byArdAweaf:-Khu6ix5dUy_srsvaOT3:bxocan8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L100)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

mentioned this elsewhere, let's go ahead and simplify this now? being able to do this refactor confidently is the reason we use TS.
</blockquote></details>

This is getting ripped out before 1.0 - the entire alert and the "help us" section.

---

*[pkg/ui/src/redux/alerts.ts, line 121 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-Khhaeau8QEFTy44keiF:-Khu6vcjuS8X9juC8Sv3:b-9u1wx4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L121)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is this nil guard necessary? is build info nullable?
</blockquote></details>

I mean, maybe?  I've seen weirder things, this is coming right from the server.

---

*[pkg/ui/src/redux/alerts.ts, line 132 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KhhalHQf3_soElL4OCY:-Khu7VANGI1QTrSGiPJn:b-trfozt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L132)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

same question about strict null checks
</blockquote></details>

I don't know what you mean, can't we just change the return of the function to `Alert | undefined` in that case?

---

*[pkg/ui/src/redux/alerts.ts, line 135 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KhhanbkuKHu6KnKq9uq:-Khu7iEkyNGDRYMJYqjY:b1s717c) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L135)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

how can versions be null or undefined?
</blockquote></details>

if NodeStatuses is null or undefined.

---

*[pkg/ui/src/redux/alerts.ts, line 151 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KhhaxFBuvwVeURsXjQJ:-Khu8mSYl73m1cntmC99:b-5usvz6) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L151)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

same question about why this is needed in addition to the below, and also this probably needs the same TODO as the above corollary.
</blockquote></details>

I mean, I think it's explained in the comments, there's also a comment in uiData, it's just a weird pattern that needs to be made explicit, this is how it works now.

---

*[pkg/ui/src/redux/alerts.ts, line 183 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-Khhb9588oUj8cdSVeQU:-Khu86l98esyIf-ighMY:ban1cx1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L183)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

sketchy null checks i think
</blockquote></details>

Yeah, missing a check, added.

---

*[pkg/ui/src/redux/alerts.ts, line 199 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KhhbCx-afv3mPizQX8T:-Khu90SHW8GlAXz22her:b-guv7xd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L199)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

these repeat a bunch, are they useful? i'd remove them all.
</blockquote></details>

I find them useful.

---

*[pkg/ui/src/redux/alerts.ts, line 227 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KhhyOSpnd8hydS1gOGn:-KhuEjw5Qxg9h3NQm9GE:b4y4zmd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L227)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think you can `return _.head(builds);`
</blockquote></details>

No, needs length to be *exactly* one.

---

*[pkg/ui/src/redux/alerts.ts, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KhhyY9m-A87X4aG3gDA:-Khu9Lp6rfjeVop5KtO9:bvbbc87) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L245)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this can do better than any, can't it? at least, `typeof AdminUIState.uiData`.
</blockquote></details>

It doesn't matter what it is, we're always comparing the same thing to the same thing *later*, which will always have the same type from typescript's point of view.

But, I guess adding a type does enforce that it's not accidentally compared to something else. Changed.

---

*[pkg/ui/src/redux/alerts.ts, line 256 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-Khhzkz0nRX4Ha-FJdPB:-KhuBM2SHzrr4sLcqbGk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L256)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`return !(_.has(uiData, key) || isInFlight(state, key))`?
</blockquote></details>

Done.

---

*[pkg/ui/src/redux/apiReducers.ts, line 29 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheDK3ZSF9kfBu6gxIc:-Khu9nEuzZinKcMfT5mI:bz8413b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/apiReducers.ts#L29)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

remove this now that the whole reducer is exported?
</blockquote></details>

"refresher-only" is used in a lot of places, "Entire reducer" is only used in tests.

---

*[pkg/ui/src/redux/apiReducers.ts, line 35 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheDOJ0SpUMwOj-jgyD:-Khu9ybCbxzbDNFeO_4f:bz8413b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/apiReducers.ts#L35)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

"refresher-only" is used in a lot of places, "Entire reducer" is only used in tests.

---

*[pkg/ui/src/redux/uiData.ts, line 34 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheBsJ9KkHQPS90FQ8k:-Khu9zeIxob7aJgyizhM:bgwr8cg) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/uiData.ts#L34)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

any reason not to make it a boolean now?
</blockquote></details>

well its persisted to the server, also we're ripping it out so it doesn't matter.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-:-Khu0hCIVQLq1y6CPxuJ:b-3sv830)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r2.
Review status: 25 of 28 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/ui/src/redux/alerts.ts, line 75 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-Khh_zPGgSaC2amTEPJI:-KhvWKH5T6R0BlUpkhez:bjgn55t) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L75)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

I think the function would just have to return Alert | undefined, we could easily switch all of these to null.  Is that what you mean?
</blockquote></details>

I was just asking; `null` would also not be allowed here because the function would then have to return `Alert?`. This is fine for now, I guess, but we should turn on strict null checks soon after 1.0.

---

*[pkg/ui/src/redux/alerts.ts, line 121 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-Khhaeau8QEFTy44keiF:-KhvWhzwlX2UaI87nhwb:b-a3vz5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L121)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

I mean, maybe?  I've seen weirder things, this is coming right from the server.
</blockquote></details>

It doesn't matter if it comes from the server; what does the typing say?

---

*[pkg/ui/src/redux/apiReducers.ts, line 29 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheDK3ZSF9kfBu6gxIc:-KhvXK3DAzHPyLSRhffl:b8dq7b2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/apiReducers.ts#L29)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

"refresher-only" is used in a lot of places, "Entire reducer" is only used in tests.
</blockquote></details>

A comment would make that obvious.

---

*[pkg/ui/src/redux/apiReducers.ts, line 35 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheDOJ0SpUMwOj-jgyD:-KhvXMKiSUK1eauA3m-D:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/apiReducers.ts#L35)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

"refresher-only" is used in a lot of places, "Entire reducer" is only used in tests.
</blockquote></details>

ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 24 of 28 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/ui/src/redux/alerts.ts, line 121 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-Khhaeau8QEFTy44keiF:-Khwvvr9rYCFUZfTCp4W:bgp7e3n) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L121)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

It doesn't matter if it comes from the server; what does the typing say?
</blockquote></details>

This is a protobuf, so *every* field is nullable, and that can be verified from the typings in proto.d.ts.

---

*[pkg/ui/src/redux/apiReducers.ts, line 29 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheDK3ZSF9kfBu6gxIc:-KhwxhTQuwoMJXuxlC5x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/apiReducers.ts#L29)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

A comment would make that obvious.
</blockquote></details>

Done.

---

*[pkg/ui/src/redux/apiReducers.ts, line 35 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-KheDOJ0SpUMwOj-jgyD:-KhwxhoMgOIQ_HjP9Twc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/apiReducers.ts#L35)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r3.
Review status: 25 of 28 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/ui/src/redux/alerts.ts, line 121 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915#-Khhaeau8QEFTy44keiF:-Khx-INK0bmKDyx6EDjy:b-hftfcd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f9a1cc9e24aa285eeab0cc1934af825df38deb1/pkg/ui/src/redux/alerts.ts#L121)):*
> This is a protobuf, so every field is nullable

That's not true of primitive fields in proto3. Thanks for checking the typings.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14915)*
<!-- Sent from Reviewable.io -->
